File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        res = dict()
        left =0 
        max_length = 0

        for right in range(len(s)):

            if s[right] in res  :
                left = max(left,res[s[right]]+1)
            
            res[s[right]] = right
            max_length = max(max_length, right - left +1)

        return max_length
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = 0;
        int left = 0;

        Map<Character, Integer> window = new HashMap<>();

        for(int right=0; right<s.length(); right++){
            char ch = s.charAt(right);

            if(window.containsKey(ch)){
                left = Math.max(left, window.get(ch)+1);
            }
            window.put(ch,  right);

            len = Math.max(len, right-left+1);
        }

        return len;
    }
}